### PR TITLE
Add basic GraphQL TMDE types and resolvers

### PR DESCRIPTION
**Pull request by SeaverDahlgren**
### Summary

Adds basic Schema fields with simple resolver functions to the `schema.go` file. This includes all fields except container tags, task tags, networks, and stats. It introduces the `taskType` and `containerType` objects. The two queries currently implemented are `Container` and `Task` which resolve to the container querying the endpoint and the task that container is running, respectively.
### Implementation details

The current container and task are fetched while serving the request then passed to the `http.request` in context. The `schema.go` file defines `taskType` and `containerType`. The resolver functions are all defined for each field in each type. Note that if a field is named exactly the same as defined by the corresponding struct (i.e. `apitask.Task` and `apicontainer.DockerContainer`), the resolver considered "trivial" and do not need to be defined. 

A custom GraphQL scalar type, `JSON` was also created in order to resolve fields like `Labels` because it resolves to a mapping. Note that mappings are not a default GraphQL scalar type.
### Testing

A new `schema_test.go` file was created to test individual resolvers. Additional test cases were also added to `task_server_setup_test.go` as well.

New tests cover the changes: 
yes
### Description for the changelog

Add basic GraphQL TMDE types and resolvers

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/handlers/gql/schema.go [hunk 1]
```diff
@@ -32,11 +46,175 @@ func CreateSchema(
...
+				if !ok {
+					return nil, errors.Errorf("GraphQL Schema Error: Could not cast to container")
+				}
+
+				return dockerContainer, nil
+			},
+		},
+		"Task": &graphql.Field{
+			Type: taskType,
+			Resolve: func(p graphql.ResolveParams) (interface{}, error) {
```
Comment (mssrivas) [inline agent/handlers/gql/schema.go:210]: Anythin thats doing extended logic like this makes more sense to be tucked as a named function

File: agent/handlers/gql/schema.go [hunk 2]
```diff
@@ -32,11 +46,175 @@ func CreateSchema(
...
+				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
+					return availabilityZone, nil
+				},
+			},
+			"LaunchType": &graphql.Field{
+				Type: graphql.String,
+			},
+			"Containers": &graphql.Field{
+				Type: graphql.NewList(containerType),
+				Resolve: func(p graphql.ResolveParams) (interface{}, error) {
```
[the remark this answers is not in the document]
Comment (SeaverDahlgren) [reply agent/handlers/gql/schema.go:176]: The issue is that this resolver requires `dockerstate` . If I create a named function, I can no longer pass in the dockerstate.

[the remark this answers is not in the document]
Comment (angelcar) [reply agent/handlers/gql/schema.go:176]: you can do something like:

```golang
Resolve: containersResolver(state),
.
.
.
func containersResolver(state dockerstate.TaskEngineState) func(p graphql.ResolveParams) (interface{}, error) {
    return func(p graphql.ResolveParams) (interface{}, error) {
        <your original resolver code here>
    }
}
```

File: agent/handlers/gql/schema.go [hunk 3]
```diff
@@ -45,11 +223,236 @@ func CreateSchema(
...
+	if dockerContainer, ok := p.Source.(*apicontainer.DockerContainer); ok {
+		return dockerContainer.Container.GetKnownExitCode(), nil
+	}
+	return nil, nil
+}
 
-	return dockerID, nil
+func containerCreatedAtResolver(p graphql.ResolveParams) (interface{}, error) {
+	if dockerContainer, ok := p.Source.(*apicontainer.DockerContainer); ok {
+		return dockerContainer.Container.GetCreatedAt().UTC().Format(time.RFC3339Nano), nil
```
[the remark this answers is not in the document]
Comment (SeaverDahlgren) [reply agent/handlers/gql/schema.go:298]: Currently, in V2, V3, V4, the time fields are returned by marshaling a time.time pointer. The default format for this is the RFC format you see here. This is to stay consistent with the previous versions.

File: agent/handlers/gql/schema_test.go [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+		DockerID:   containerID,
+		DockerName: containerName,
+		Container:  container,
+	}
+	containerNameToDockerContainer := map[string]*apicontainer.DockerContainer{
+		taskARN: dockerContainer,
+	}
+	if containerNameToDockerContainer == nil {
+		return
+	}
```
[the remark this answers is not in the document]
Comment (SeaverDahlgren) [reply agent/handlers/gql/schema_test.go:216]: It's not :)

File: agent/handlers/gql/schema.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (SeaverDahlgren) [reply agent/handlers/gql/schema.go:176]: So I looked at what you suggested and the problem is the `Resolve` parameter takes in the type `FieldResolveFn` which is defined: `type FieldResolveFn func(p ResolveParams) (interface{}, error)`, so the only argument can be ResolveParams. I am getting this error:

`Cannot use 'taskContainersResolver(stats)' (type (interface{}, error)) as the type FieldResolveFn`

**Review by angelcar — DISMISSED** [review]

File: agent/handlers/gql/schema.go [hunk 5]
```diff
@@ -15,11 +15,21 @@ package gql
...
 import (
 	"github.com/aws/amazon-ecs-agent/agent/api"
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitask "github.com/aws/amazon-ecs-agent/agent/api/task"
 	"github.com/aws/amazon-ecs-agent/agent/engine/dockerstate"
 	"github.com/aws/amazon-ecs-agent/agent/handlers/utils"
+	v2 "github.com/aws/amazon-ecs-agent/agent/handlers/v2"
 	v3 "github.com/aws/amazon-ecs-agent/agent/handlers/v3"
+
+	"time"
```
Comment (sharanyad) [inline agent/handlers/gql/schema.go:25]: can you move this as the first import, generally the import order is - 

go std packages 

ecs agent packages 

external/third party

**Review by sharanyad — APPROVED** [review]

**Review by mssrivas — APPROVED** [review]

**Review by angelcar — APPROVED** [review]

**Review by mssrivas — APPROVED** [review]

**Pull request merged**